### DISCUSSION TITLE
Issue: Multiple fields with multiple filters loosing filters in DocumentState

### DISCUSSION THREAD

**Discussion by aj-swipejobs** [MorphiaOrg/morphia#2818] [Q&A]
so when we have multiple filters for the same property we are loosing the first filter in the map. I have tracked it down to the class "DocumentState" method "andTogether".

eg. is let's say we have following filters.
"attrib_A": "$exists"
"attrib_A": "$ne: ''"
"attrib_B": "$exists"
"attrib_B": "$ne: ''"

Document state would loose "attrib_B": "$exists" in the "finished" map.

`
private Document andTogether(Document doc, String key, @Nullable Object additional) {

        if (additional != null) {

            Document newSubdoc = new Document(key, additional);

            var extant = doc.remove(key);

            List<Document> and = (List<Document>) doc.get("$and");

            if (and != null) {

                and.add(newSubdoc);

            } else {

                and = new ArrayList<>();

                and.addAll(of(new Document(key, extant), newSubdoc));

                doc.put("$and", and);

                return newSubdoc;

            }

        }

        return doc;

    }`

in this code. 
first filter
finished = {"attrib_A" : "$exists"}
second filter
finished = {"$and": ["attrib_A" : "$exists", "attrib_A" : "$ne: ''"]}
third filter
finished = {"$and": ["attrib_A" : "$exists", "attrib_A" : "$ne: ''"], "attrib_B": "$exists"}
fourth filter
'var extant = doc.remove(key);' this would remove "attrib_B": "$exists" and "if (and != null)" would be true so 
finished = {"$and": ["attrib_A" : "$exists", "attrib_A" : "$ne: ''", "attrib_B": "$ne: ''"],}

Is this a bug or how it's suppose to work?

**Comment by evanchooly** [comment]
I would need to see your actual code but this looks like a big fixed in 2.4.11. If you could try that version to be sure the problem persists, we can dig a bit deeper.

**Reply by aj-swipejobs** [reply]
Okay, so My code looks like as following. (Sorry I forgot to add, I'm using morphia version 2.4.8)


```
{
List<Filter> filters = new ArrayList<>();
        filters.add(eq(ID, subOrderId));
        filters.add(gte(START_TIME,rangeStart));
        filters.add(exists(GROUP_ID));
        filters.add(ne(GROUP_ID, ""));
        if (rangeEnd != null) {
            filters.add(lt(START_TIME, rangeEnd));
        }
        filters.add(eq(STATUS, VISIBLE));

       return getCursorBy(filters, new FindOptions()).toList();
}
        
        public MorphiaCursor<T> getCursorBy(List<Filter> filters, FindOptions findOptions) {
        Filter[] filterArray = filters.toArray(new Filter[]{});

        return getQuery()
                .filter(filterArray)
                .iterator(findOptions);
    }
    
        protected Query<T> getQuery(){

        return this.datastore.find(genericType);
    }
```

**Reply by aj-swipejobs** [reply]
Meanwhile I'll try it with 2.4.11, Thanks 👍

**Reply by aj-swipejobs** [reply]
Hi @evanchooly I have tested using version 2.4.11 well, it kind of works now. so with Filters.eq it fails `current instanceof Document` in MergingDocument(inner class in DocumentState) and it would get replaced. 
Following filter list would only contain aString: {$ne: null}
```
List<Filter> filters = of(
                gte("number", NUMBER_1),
                eq("aString", TEST_STRING_2),
                ne("aString", null),
                lte("number", NUMBER_5)
        );
```

but if we shuffle it around like
```
List<Filter> filters = of(
                gte("number", NUMBER_1),
                ne("aString", null),
                lte("number", NUMBER_5),
                eq("aString", TEST_STRING_2)
        );
```
then query would contain both aString: {$ne: null} and aString: "TEST_STRING_2"

Please let me know if you need more information regarding this

**Reply by evanchooly** [reply]
Can you attach a stack trace? If not it should be easy enough to recreate but maybe it'll save some time. I'm prepping for a talk tomorrow night but I'd that goes well, I'll try to give it a look tonight.

**Reply by aj-swipejobs** [reply]
Hi, @evanchooly it doesn't throw any errors, code just works but it's missing the eq filter when it runs the query.

**Reply by gnewitt** [reply]
To be fair Evan, my example above is perhaps a bit nonsensical and that may be why it fails.
The query does an eq() filter and then a ne(null) filter.  Now I guess being logical, for a given field eq() will be true for a given result and any other filters should be unncessary.  If that is the upshot here, then fair enough.

I guess if I rearrange the filters, I will get my answer on that score.

**Reply by gnewitt** [reply]
Having done this:

`    void test_multiplefields_multiplefilters() {
        List<Filter> filters = of(
                eq("aString", TEST_STRING_2),
                ne("aString", null),
                gt("number", NUMBER_1),
                lt("number", NUMBER_5)
        );
        List<TestClass> results = testRepository.getListBy(filters);

        assertThat(results, is(notNullValue()));
        assertThat(results.isEmpty(), is(false));
        assertThat(results.size(), is(2));
    }
`

I see this issue still manifested, such that I believe the gt() filter is ignored.  This is using 2.4.11

Cheers,

Greg

**Reply by evanchooly** [reply]
this is all good to know.  within reason, morphia should replicate your query structure regardless of how nonsensical it may be.  let the server handle that but morphia should just transcribe what it's given.  I'll try to tackle this tomorrow night (i'm presenting at a JUG tonight so I won't be home until late).

**Reply by gnewitt** [reply]
Delving deeper, the issue appears to be when there is an eq() filter for a particular field and then another filter for the same field is added.  So as in the example above, if I set up an eq() filter and then follow it with another filter (in this case ne()), the eq() filter seems to be disregarded.
